util/tracing/tracer_span.go

```diff
@@ -0,0 +1,568 @@
+	TraceID uint64
+
+	// A probabilistically unique identifier for a span.
+	SpanID uint64
+}
+
+type spanContext struct {
+	spanMeta
+
+	shadowTr  *shadowTracer
+	shadowCtx opentracing.SpanContext
+
+	// If set, all spans derived from this context are being recorded as a group.
+	recordingGroup *spanGroup
+	recordingType  RecordingType
+
+	// The span's associated baggage.
+	Baggage map[string]string
+}
+
+const (
+	// TagPrefix is prefixed to all tags that should be output in SHOW TRACE.
+	TagPrefix = "tidb."
+	// StatTagPrefix is prefixed to all stats output in span tags.
+	StatTagPrefix = TagPrefix + "stat."
+)
+
+// SpanStats are stats that can be added to a span.
+type SpanStats interface {
+	proto.Message
+	// Stats returns the stats that the object represents as a map from stat name
+	// to value to be added to span tags. The keys will be prefixed with
+	// StatTagPrefix.
+	Stats() map[string]string
+}
+
+var _ opentracing.SpanContext = &spanContext{}
+
+// ForeachBaggageItem is part of the opentracing.SpanContext interface.
+func (sc *spanContext) ForeachBaggageItem(handler func(k, v string) bool) {
+	for k, v := range sc.Baggage {
+		if !handler(k, v) {
+			break
+		}
+	}
+}
+
+// RecordingType is the type of recording that a span might be performing.
+type RecordingType int
+
+const (
+	// NoRecording means that the span isn't recording.
+	NoRecording RecordingType = iota
+	// TiDBRecording means that remote child spans (generally opened through
+	// RPCs) are also recorded.
+	TiDBRecording
+	// SingleNodeRecording means that only spans on the current node are recorded.
+	SingleNodeRecording
+)
+
+type span struct {
+	spanMeta
+
+	parentSpanID uint64
+
+	tracer *Tracer
+
+	// x/net/trace.Trace instance; nil if not tracing to x/net/trace.
+	netTr trace.Trace
+	// Shadow tracer and span; nil if not using a shadow tracer.
+	shadowTr   *shadowTracer
+	shadowSpan opentracing.Span
+
+	operation string
+	startTime time.Time
+
+	// Atomic flag used to avoid taking the mutex in the hot path.
+	recording int32
+
+	mu struct {
+		sync.RWMutex
+		// duration is initialized to -1 and set on Finish().
+		duration time.Duration
+
+		recordingGroup *spanGroup
+		recordingType  RecordingType
+		recordedLogs   []opentracing.LogRecord
+		// tags are only set when recording.
+		tags opentracing.Tags
+
+		stats SpanStats
+
+		// The span's associated baggage.
+		Baggage map[string]string
+	}
+}
+
+var _ opentracing.Span = &span{}
+
+func (s *span) isRecording() bool {
+	return atomic.LoadInt32(&s.recording) != 0
+}
+
+// IsRecording returns true if the span is recording its events.
+func IsRecording(s opentracing.Span) bool {
+	if _, noop := s.(*noopSpan); noop {
+		return false
+	}
+	return s.(*span).isRecording()
+}
+
+func (s *span) enableRecording(group *spanGroup, recType RecordingType) {
+	if group == nil {
+		panic("no spanGroup")
+	}
+	s.mu.Lock()
+	atomic.StoreInt32(&s.recording, 1)
```

Comment:
already locked does we need atomic write operation here in golang? I'm not confirm.. @tiancaiamao